cmd/zrepl/zrepl.c

```diff
@@ -513,6 +522,53 @@ uzfs_zvol_mgmt_do_handshake(zvol_io_hdr_t *hdr, int sfd, char *name)
 	return (rc);
 }
 
+static int
+uzfs_zvol_connect_to_tgt_controller(void *arg)
+{
+	char ip_buf[256];
+	int sfd, rc;
+	struct sockaddr_in istgt_addr;
+	const char *target_addr = arg;
+
+	if (target_addr == NULL) {
+		if (get_controller_ip_address(ip_buf, sizeof (ip_buf)) != 0) {
+			printf("parsing IP address did not work\n");
+			return (-1);
+		}
+		target_addr = ip_buf;
+	}
+
+	printf("Controller IP address is: %s\n", target_addr);
+	sfd = create_and_bind(mgmt_port, false);
+	if (sfd == -1) {
+		return (-1);
+	}
+
+	rc = make_socket_non_blocking(sfd);
+	if (rc == -1) {
+		return (-1);
+	}
+
+	bzero((char *)&istgt_addr, sizeof (istgt_addr));
+	istgt_addr.sin_family = AF_INET;
+	istgt_addr.sin_addr.s_addr = inet_addr(target_addr);
+	istgt_addr.sin_port = htons(TARGET_PORT);
+retry:
+	rc = connect(sfd, (struct sockaddr *)&istgt_addr, sizeof (istgt_addr));
+	if ((rc == -1) && ((errno == EINTR) || (errno == ECONNREFUSED) ||
```

Comment:
rc being -1 is itself error.. errnos can be used for retry logic

Reply:
Thats what this code doing. It check of return value and then check errno to decide if it need to retry or not. Need code re-orgainzation for readability ?

Reply:
not for readability @satbirchhikara .. lets say rc is -1 and errno is something other than EINTR/ECONNREFUSED/ETIMEDOUT/EINPROGRESS, its taken as success

Reply:
also EINPROGRESS probably need to be handled separately as same connection can get succeeded sometime later due to non-blocking nature..
```       EINPROGRESS
              The  socket is nonblocking and the connection cannot be completed immediately.  It is possible to select(2) or poll(2) for com‐
              pletion by selecting the socket for writing.  After select(2) indicates writability, use getsockopt(2)  to  read  the  SO_ERROR
              option at level SOL_SOCKET to determine whether connect() completed successfully (SO_ERROR is zero) or unsuccessfully (SO_ERROR
              is one of the usual error codes listed here, explaining the reason for the failure).
```

Reply:
You have point here.